src/work_queue.cpp

```diff
@@ -102,6 +102,7 @@ bool WorkQueue::peek_batch(std::vector<DataNode *> &node_vec, int batch_size) {
 
   // wait until consumer queue is large enough
   std::unique_lock<std::mutex> lk(consumer_list_lock);
+  //FIXME: This version of batch waiting causes a lot of contention on the mutex.
```

Comment:
How so?

Reply:
Essentially, when batch waiting, every single insert will wake up this thread. This thread will take the mutex, then go back to sleep. Taking this mutex can slow down things being inserted in as they need the mutex too.

The solution would be to only wake up when there's a batch_waiting_target threshold crossed.